### PR TITLE
authenticate with unverified email and allowUnverified

### LINKED ISSUE #2902: [b2g][unverified] fatal error when trying to reuse unverified email address
STR

1) Using a current unagi build on a phone, configure /data/local/user.js to enable native navigator.id:

```
pref("dom.identity.enabled", true);
pref("toolkit.identity.debug", true);
pref("layout.css.report_errors", false);
```

(The layout.css.report_errors isn't necessary, but seriously, the noise about CSS "declaration dropped" is useless information).
2) Use the firefox browser on the phone and go to `https://marketplace-dev.allizom.org`
3) Click on `Log in/Register` to bring up the trusted UI.
4) Enter an email address that has not been used before; then enter a password twice. You will be authenticated.
5) (Also an email to confirm that address will be sent out; do not verify that email address).
6) Now logout of marketplace-dev.
7) Click `Log in/Register` to bring up the dialog again.
8) Click `This is not me`. Re-enter the same email address from step 4) and a password twice.

Expected: this unverified user should be restaged again

Actual: "We are very sorry. The server is under extreme load."

```
Action: Creating Account
Network Info: POST /wsapi/stage_user
Response Code: 503
Response Text/Error Type: Service Unavailable: database unavailable
```

In the dbwriter.log on notoriousb2g.personatest.org:

```
unexpected database failure: Error: Duplicate entry 'blahblah@gmail.com' for key 'address' -- at exports.createUnverifiedUser
(/home/app/code/lib/db/mysql.js:680:13)
```

### PR DESCRIPTION
**Pull request by seanmonstar**
fixes #2902

### CODE REVIEW

File: resources/static/dialog/js/modules/authenticate.js [hunk 1]
```diff
@@ -85,13 +88,13 @@ BrowserID.Modules.Authenticate = (function() {
         self.close("primary_user", info, info);
       }
       else if (!!self.forceIssuer && 'default' !== self.forceIssuer) {
-	if (hasPassword(info)) {
+        if (hasPassword.call(self, info)) {
           enterPasswordState.call(self);
-	} else {
+        } else {
```
Comment (ozten) [inline resources/static/dialog/js/modules/authenticate.js:93]: Thanks for the cleanup.

File: resources/static/dialog/js/modules/authenticate.js [hunk 2]
```diff
@@ -39,15 +39,18 @@ BrowserID.Modules.Authenticate = (function() {
...
+       info.state === "transition_to_secondary" ||
+       info.state === "unverified" && this.allowUnverified));
   }
 
   function initialState(info) {
     /*jshint validthis: true*/
     var self=this;
     self.submit = checkEmail;
-    if (hasPassword(info)) {
+    if (hasPassword.call(self, info)) {
```
Comment (ozten) [inline resources/static/dialog/js/modules/authenticate.js:53]: Ah, good catch.

File: resources/static/test/cases/dialog/js/modules/authenticate.js [hunk 3]
```diff
@@ -114,6 +114,21 @@
     });
   });
 
+  asyncTest("allowUnverified with an unverified email declared in options - show password field", function() {
+    controller.destroy();
```
Comment (ozten) [inline resources/static/test/cases/dialog/js/modules/authenticate.js:118]: Excellent.

File: resources/static/dialog/js/modules/authenticate.js (hunk 2 above)
Comment (seanmonstar) [inline resources/static/dialog/js/modules/authenticate.js:53]: It wasn't a bug before. I just needed access to `this` inside hasPassword, to check for `allowUnverified`

**Pull request merged**